metrics/bigquery.py

```diff
@@ -50,67 +50,159 @@ def validate_metric_name(name):
     if not re.match(r'^[\w-]+$', name) or name[-1] == '\n':
         raise ValueError(name)
 
-def do_query(query, out_filename, project):
-    """Executes a bigquery query, outputting the results to a file."""
+def do_jq(jq_filter, data_filename, out_filename, jq_bin='jq'):
+    """Executes jq on a file and outputs the results to a file."""
     with open(out_filename, 'w') as out_file:
-        cmd = [
-            'bq', 'query', '--format=prettyjson', '--project_id=%s' % project,
-            '-n100000',  # Results may have more than 100 rows
-        ]
-        check(cmd, stdin=query, stdout=out_file)
-
-def do_jq(jqfilter, data_filename, out_filename, jq_bin='jq'):
-    """Executes jq on a data file and outputs the results to a file."""
-    with open(out_filename, 'w') as out_file:
-        check([jq_bin, jqfilter, data_filename], stdout=out_file)
-
-def run_metric(project, prefix, metric, query, jqfilter, jq_point=None, influx=None):
-    """Runs query and filters results, uploading data to GCS."""
-    stamp = time.strftime('%Y-%m-%d')
-    raw = 'raw-%s.json' % stamp
-    filtered = 'daily-%s.json' % stamp
-    latest = '%s-latest.json' % metric
-    points = '%s-data-points.json' % metric
-
-    do_query(query, raw, project)
-    do_jq(jqfilter, raw, filtered)
-    if jq_point:
-        if influx:
-            do_jq(jq_point, raw, points)
-            with open(points) as points_file:
-                points = json.load(
-                    points_file,
-                    object_hook=influxdb.Point.from_dict,
+        check([jq_bin, jq_filter, data_filename], stdout=out_file)
+
+class BigQueryer(object):
+    def __init__(self, project, bucket_path, backfill_days, influx_client):
+        if not project:
+            raise ValueError('project', project)
+        self.project = project
+        if not bucket_path:
+            raise ValueError('bucket_path', bucket_path)
+        self.prefix = bucket_path
+
+        self.influx = influx_client or None
+        self.backfill_days = backfill_days
+        # These functions take some action on query results that are produced.
+        self.result_consumers = [self.jq_upload, self.influx_upload]
+
+    def do_query(self, query, out_filename):
+        """Executes a bigquery query, outputting the results to a file."""
+        with open(out_filename, 'w') as out_file:
+            cmd = [
```

Comment:
create `cmd` outside the `with` statement

Reply:
done